packages/contracts/src/core/permission/PermissionManager.sol

```diff
@@ -109,6 +182,213 @@ abstract contract PermissionManager is Initializable {
+        Permission storage permission = permissions[permHash];
+
+        if (_isPermissionFrozen(permission)) {
+            revert PermissionFrozen(_where, _permissionIdOrSelector);
+        }
+
+        if (!hasPermission(permission.owners[msg.sender], _flags)) {
+            revert InvalidOwnerPermission(msg.sender, permission.owners[msg.sender], _flags);
+        }
+
+        uint8 newFlags = permission.delegations[_delegatee][permHash] | _flags;
+        permission.delegations[_delegatee][permHash] = newFlags;
+
+        emit PermissionDelegated(_where, _permissionIdOrSelector, _delegatee, newFlags);
+    }
+
+    /// @notice Function to remove sepcific flags from the delegatee
+    /// @param _where The address of the target contract for which `_who` receives permission.
+    /// @param _permissionIdOrSelector The permission hash or function selector used for this permission.
+    /// @param _delegatee The addresses we want to undelegate specifc flags.
+    /// @param _flags The flags as uint8 the permission owner wants to remove from this specific delegatee.
+    function undelegatePermission(
+        address _where,
+        bytes32 _permissionIdOrSelector,
+        address _delegatee,
+        uint8 _flags
+    ) public {
+        bytes32 permHash = permissionHash(_where, _permissionIdOrSelector);
+        Permission storage permission = permissions[permHash];
+
+        if (!hasPermission(permission.owners[msg.sender], _flags)) {
+            revert InvalidOwnerPermission(msg.sender, permission.owners[msg.sender], _flags);
+        }
+
+        uint8 newFlags = permission.delegations[_delegatee][permHash] ^ _flags;
+        permission.delegations[_delegatee][permHash] = newFlags;
+
+        emit PermissionUndelegated(_where, _permissionIdOrSelector, _delegatee, newFlags);
+    }
+
+    /// @notice Function to add a new owner to a permission.
+    /// @param _where The address of the target contract for which `_who` receives permission.
+    /// @param _permissionIdOrSelector The permission hash or function selector used for this permission.
+    /// @param _owner The new manager for this permission.
+    /// @param _flags The flags as uint8 to restrict what this specifc owner actually can do. (only revoke? only grant? both?)
+    function addOwner(
+        address _where,
+        bytes32 _permissionIdOrSelector,
+        address _owner,
+        uint8 _flags
+    ) external {
+        if (_owner == address(0)) {
+            revert ZeroAddress();
+        }
+
+        Permission storage permission = permissions[permissionHash(_where, _permissionIdOrSelector)];
+
+        if (_isPermissionFrozen(permission)) {
+            revert PermissionFrozen(_where, _permissionIdOrSelector);
+        }
+
+        if (!hasPermission(permission.owners[msg.sender], _flags)) {
+            revert InvalidOwnerPermission(msg.sender, permission.owners[msg.sender], _flags);
+        }
+
+        uint8 currentFlags = permission.owners[_owner];
+
```

Comment:
`_owner` might already be the owner with a flag but you still emit the event in the end.

I think either we should revert if owner is already added or silently succeed, but not emit the event.

Same problem in `removeOwner`.

Reply:
I think we need to emit the event. Based on the flags in the event, the indexer can create a good history of the flag changes.

Reply:
Yes, true but what I meant was: 

assume owner already exists with `grantOwner` flag and someone adds him again with the same `grantOwner` flag.  Shall we still emit the event ? as it really doesn't make any sense for the history

Reply:
Yeah, this we can filter out in the indexer.. if even needed to do so.. eg for a user it could be great to see that he accidentally passed the same flags instead of the one that would extend the permissions..